python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+minimum viable implementation of the PACE Python interface,
+and also the prototype Python class wrapper intended to be used by the PACE Matlab interface.
+
+We first detail how Matlab classes are wrapped in a Python proxy class which exposes their properties and methods to Python.
+We then move to how communications between PACE programs written in different languages is facilitated.
+Finally we describe remaining issues with the approach taken and how the prototype may be improved for production.
+
+
+# Compiled Matlab
+
+As noted in [ADR #10](../../documentation/adr/0010-use-compiled-matlab-for-python.md) the decision was taken
+to use "compiled" Matlab code for the Python interface to Horace/Herbert rather than to rewrite them in Python/C++.
+This imposes certain constraints on an "pyHorace" implementation:
+
+1. Only user written "gateway" m-file *functions* maybe directly accessed by Python (or executed on the commandline) in a compiled Matlab code.
+2. Matlab is run as a loaded library within the Python process ("in-process" execution).
+3. The Matlab help system is not available to compiled Matlab code.
+4. The Matlab MCR requires the use of its own standard libraries which implies restrictions on which compilers/Python versions can be used.
+
+Point 1 means that Matlab classes and standard Matlab functions cannot be *directly* created or called from Python 
+(a specific gateway function which constructs the class must be used instead).
+However, any Matlab code (and indeed the entire standard Matlab built-in library of functions and classes) maybe in "included" in the compiled library.
+The restrictions is that this "included" code may only be accessed by the user-written m-file *function* (not class)
+and not by the user calling those functions/classes from Python.
+Given this major limitation, the approach taken by the "Hugo" prototype is to use a gateway function,
+`call_method` which takes the *name* of a Matlab function, class or class method and a set of arguments and uses
+`feval` to call those "included" functions/class/methods.
+This is described in more detail in the [next section](#gateway-functions-and-proxy-classes).
+
+Point 2 means that the native Matlab way of calling Python (using the `py.*` namespace) cannot be used
+but also means that a `mex` file can be used for this purpose.
+This is described in detail in the [section on Python user functions](#calling-python-from-compiled-matlab).
+
+Point 3 means that if we would like users to be able to use `help(sqw)` or similar
+we would need to generate the help text and wrappers for each function / class we export (not currently implemented).
+This and point 4 and other limitations are discussed in the [limitations section](#limitations-and-future-work)
+
+
+# Gateway functions and proxy classes
+
+["pyHorace"](https://github.com/mducle/hugo) is based on a framework (["pySpinW"](https://github.com/spinw/pyspinw)) written for SpinW by S. Ward.
```

Comment:
This link references "Hugo" (the prototype) but talks of "pyHorace". The naming is unclear and the first link is simply to "prototype"

Reply:
Yeah... we need to decide on name... "Hugo" is just my whimsiness... the actual syntax uses `pyhorace`. But I think it should be called `pypace` or `pace-python` or some such...